### PR TITLE
2D metallicity model

### PR DESCRIPTION
**Pull request by PabloCorcho**
Implementation of a new metallicity 2D interpolation method for Tabula_MFH models.

### CODE REVIEW

**Comment by PabloCorcho** [comment]
Right now, this PR only includes a proof of concept

File: src/pst/models.py [hunk 1]
```diff
@@ -22,11 +24,22 @@ class ChemicalEvolutionModel(ABC):
     -------
 
     """
+    
     def __init__(self, **kwargs):
-        self.M_gas = kwargs.get('M_gas', 0*u.Msun)
-        self.Z = kwargs.get('Z', 0.02)
```
Comment (paranoya) [inline src/pst/models.py:27]: I'm now more inclined towards full cleaning before the release

File: src/pst/models.py [hunk 2]
```diff
@@ -47,32 +60,56 @@ def interpolate_ssp_masses(self, SSP: pst.SSP.SSPBase, t_obs: u.Quantity):
...
             [0 << u.yr, np.sqrt(SSP.ages[1:] * SSP.ages[:-1]), 1e12 << u.yr])
-        t_bins = t_obs - age_bins
-        t_bins = t_bins[t_bins >= 0]
-        M_t = self.integral_SFR(t_bins)
-        M_bin = np.hstack([M_t[:-1]-M_t[1:], M_t[-1]])
-        MZ_t = self.integral_Z_SFR(t_bins)
-        MZ_bin = np.hstack([MZ_t[:-1]-MZ_t[1:], MZ_t[-1]])
-        z_bin = np.clip(MZ_bin / (M_bin + 1 * u.kg),
-                        SSP.metallicities[0],
-                        SSP.metallicities[-1]) << u.dimensionless_unscaled
```
Comment (paranoya) [inline src/pst/models.py:60]: The new interpolation scheme is the important part

File: src/pst/models.py [hunk 3]
```diff
@@ -154,13 +191,11 @@ def compute_photometry(self, ssp, t_obs, photometry=None,
         model_photometry = np.sum(photometry * weights, axis=(-1, -2))
         return model_photometry
 
-    @abstractmethod
-    def integral_SFR(self):
-       pass
-
+    '''
```
Comment (paranoya) [inline src/pst/models.py:194]: This has been renamed to `stellar_mass_formed` in this module. I think it is only called from `tests/analytic_sfhs.py`, `tests/resolution_effects_on_tabular_sfhs.py`, and `src/pst/fitting_module.py`. All these will be cleaned up, won't they?

File: src/pst/models.py [hunk 4]
```diff
@@ -561,16 +599,43 @@ def integral_SFR(self, times: u.Quantity):
...
+        integral[times < self.table_t[0]] = self.table_metallicity[0]
 
         # integral = np.interp(times.to(self.table_t.unit).value,
         #                      self.table_t.value, self.table_M.value,
         #                      left=0, right=self.table_M[-1].value) * self.table_M.unit
         return integral
     
+    '''
     def integral_Z_SFR(self, times):
         """Evaluate the integral of the average metallicity over a given set of times.
```
Comment (paranoya) [inline src/pst/models.py:640]: Fixed this. We may delete this method, or turn it into `mass_weighted_stellar_metallicity`

File: src/pst/models.py [hunk 5]
```diff
@@ -610,7 +673,84 @@ def dot_SFR(self, times):
     def ddot_SFR(self, times):
         return np.interp(times, self.table_t, self.table_ddot_SFR, left=0, right=0)
 
+    def interpolate_ssp_masses(self, ssp: pst.SSP.SSPBase, t_obs: u.Quantity):
```
Comment (paranoya) [inline src/pst/models.py:676]: This method is provisionally commented out, prior to removal

File: src/pst/models.py [hunk 6]
```diff
@@ -610,7 +673,84 @@ def dot_SFR(self, times):
...
+        M_t = self.stellar_mass_formed(t_bins)
+        M_bin = np.hstack([M_t[:-1]-M_t[1:], M_t[-1]])
+
+        weights = np.zeros((ssp.metallicities.size, ssp.ages.size))
+        t_indices = np.arange(0, M_bin.size, dtype=int)
+        weights[:, t_indices] = z_weights * M_bin[np.newaxis, :]
+        return weights << u.Msun
+    '''
+
+'''
```
Comment (paranoya) [inline src/pst/models.py:753]: Many of the following are also commented

File: src/pst/models.py (hunk 3 above)
Comment (PabloCorcho) [reply src/pst/models.py:194]: Yes, they will.

File: src/pst/models.py [hunk 7]
```diff
@@ -49,32 +59,56 @@ def interpolate_ssp_masses(self, SSP: pst.SSP.SSPBase, t_obs: u.Quantity):
...
+        # test plot
+        fig, axes = plt.subplots(nrows=1, ncols=1)
+        ax = axes
+        ax.plot(bin_age, bin_metallicity, 'k-+')
+        ax.plot(t_obs-self.table_t, self.table_metallicity, 'm-+')
+        ax.set_xscale('log')
+        '''
+        
+        # SSP linear interpolation
+        age_index = np.clip(SSP.ages.searchsorted(bin_age, side='right'),
```
Comment (PabloCorcho) [inline src/pst/models.py:90]: @paranoya, I think there was a slightly small bug. The index was assumed to correspond to the index of the upper value, but it is the right one that is returned (see [np.searchshorted](https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html)). I have included the `side=right` to fix that.

File: src/pst/models.py [hunk 8]
```diff
@@ -49,32 +59,56 @@ def interpolate_ssp_masses(self, SSP: pst.SSP.SSPBase, t_obs: u.Quantity):
...
+        # oversample
+        w1 = np.arange(oversample_factor) / oversample_factor
+        age_bins = np.hstack(
+            [(1-w1)*age_bins[i] + w1*age_bins[i+1] for i in range(age_bins.size - 1)]
+            + [t_obs])
+
+        # find bin properties
+        mass = self.stellar_mass_formed(t_obs - age_bins).to_value(u.Msun)
+        bin_mass = mass[:-1] - mass[1:]
+        bin_age = (age_bins[1:] + age_bins[:-1]) / 2
```
Comment (PabloCorcho) [inline src/pst/models.py:77]: Why this is done in linear scale and not in log?

**Pull request merged**